### ISSUE TITLE
Dynamics and Back constraints for GPLVM

### ISSUE THREAD

**Issue by buntyke** [GPflow/GPflow#321]
I am trying to implement Gaussian Process Dynamical Model ([GPDM](https://papers.nips.cc/paper/2783-gaussian-process-dynamical-models.pdf)) using GPflow.

My [implementation](https://github.com/buntyke/GPflow/blob/master/GPflow/gpdm.py#L55) derives from GPLVM and GPR classes. The `build_likelihood()` function is based on Eqn 16 in the paper without any priors on the hyper parameters. However, the learned latent space does not seem to generate a latent space very different from GPLVM. I have evaluated its performance on two datasets in an [IPython notebook](https://github.com/buntyke/GPflow/blob/master/doc/source/notebooks/GPLVM2.ipynb) for your reference.

Could the maintainers provide feedback with the following:
1) Do you think my implementation is correct?
2) How else can I verify if my implementation is correct?
3) If the implementation is wrong, any pointers on how to fix it?

Thanks,
Nishanth

**Comment by markvdw** [comment]
I don't really have time to go over your implementation, but have you tried learning a simple linear dynamical system yet? This is the most simple dynamical system, and in low noise, should be very amenable to learning.

**Comment by buntyke** [comment]
Hi Mark,
Thanks for your feedback. Today, I implemented a simple Linear Dynamical System as follows:
* d = 2, D = 10
* Linear Dynamics without noise, x_t = diag([-1 -2]) x_{t-1}
* Linear Mapping, y_t = A x_t, A \in R^{Dxd} or Nonlinear Mapping y_t = sin(A x_t)*tanh(B x_t) with noise in both cases

I trained GPLVM and GPDM models for both scenarios and GPDM was able to learn the linear dynamics. Please check the results at [Notebook](https://github.com/buntyke/GPflow/blob/master/doc/source/notebooks/GPDM.ipynb). It looks like the implementation is working well and for real-world datasets the modelling needs to be done much more carefully.

Would you accept a pull request of these implementations? I am also planning on implementing [Variational Gaussian Process Dynamical Systems](https://papers.nips.cc/paper/4330-variational-gaussian-process-dynamical-systems.pdf)

Thanks,
Nishanth

**Comment by buntyke** [comment]
I am currently trying to implement [Variational GPDS](https://arxiv.org/pdf/1107.4985v1.pdf) by extending the `BayesianGPLVM `class. For this implementation, I need to update the KL divergence term as given in Eqn 32 by using the reparametrization trick to the variational parameters given by Eqn 15. The new KL divergence term is given by:
```
- NQ/2 -Q/2 log |K_t| - 1/2 \sum_{q=1}^Q [ |Sq| + Tr(Kt^-1 Sq) + Tr(Kt^-1 muq muq^T) ] 
Sq = [Kt^-1 + diag(lambda_q)]^-1, mu_q = Kt muhat_q
```

There is a [matlab implementation](https://github.com/SheffieldML/vargplvm/blob/c9053c2ff35d849c9c42b2856c5871b20200781d/vargplvm/matlab/vargpTimeDynamicsVarPriorBound.m#L94) of this term available in the vargplvm toolbox. However, I am not able to implement this efficiently in tensorflow and the gradients are exploding. Please see my [implementation](https://github.com/buntyke/GPflow/blob/master/GPflow/vgpds.py#L103) here.

Could the maintainers provide some pointers on how to implement this expression efficiently. For example, how can I reformulate it so that I don't explicitly compute matrix inverses and maybe use cholesky decomposition?

**Comment by buntyke** [comment]
Thank you for the reference @nknudde. I will close this thread now and move the discussion to the new issue you created.
